contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,427 @@
+     * @dev max tradeable amount increases with each block and gets reset at each trade for a token
+     */
+    function _maxTradeableAmount(Token token, uint256 tradeAmount) private view returns (uint256) {
+        uint256 lastBlockTraded = _lastTradedBlock[token];
+        // if first trade for a token, check how many blocks have been mined since deployment
+        uint256 lastBlock = lastBlockTraded == 0 ? _deploymentBlock : lastBlockTraded;
+        uint256 blocksMined = block.number - lastBlock;
+        uint256 maxPPBTradeable = _maxTradeablePPBPerBlock * blocksMined;
+        // max tradeable is capped at 100% of tradeAmount
+        if (maxPPBTradeable > PPB_RESOLUTION) {
+            maxPPBTradeable = PPB_RESOLUTION;
+        }
+        return MathEx.mulDivF(tradeAmount, maxPPBTradeable, PPB_RESOLUTION);
+    }
+
+    /**
+     * @dev sets the last traded block for a token to the current block
+     */
+    function _updateLastTradedBlock(Token token) private {
+        _lastTradedBlock[token] = block.number;
+    }
+
+    function _setRewardsPPM(uint32 newRewardsPPM) private {
+        uint32 prevRewardsPPM = _rewardsPPM;
+
+        // return if the rewards percentage PPM is the same
+        if (prevRewardsPPM == newRewardsPPM) {
+            return;
+        }
+
+        _rewardsPPM = newRewardsPPM;
+
+        emit RewardsUpdated({ prevRewardsPPM: prevRewardsPPM, newRewardsPPM: newRewardsPPM });
+    }
+
+    function _setMaxTradeablePPB(uint32 newMaxTradeablePPB) private {
+        uint32 prevMaxTradeablePPB = _maxTradeablePPBPerBlock;
+
+        // return if the max tradeable PPB is the same
+        if (prevMaxTradeablePPB == newMaxTradeablePPB) {
+            return;
+        }
+
+        _maxTradeablePPBPerBlock = newMaxTradeablePPB;
+
+        emit MaxTradeableUpdated({ prevMaxTradeablePPB: prevMaxTradeablePPB, newMaxTradeablePPB: newMaxTradeablePPB });
+    }
+
+    function _setMaxSlippagePPM(uint32 newMaxSlippagePPM) private {
+        uint32 prevMaxSlippagePPM = _maxSlippagePPM;
+
+        // return if the max slippage PPM is the same
+        if (prevMaxSlippagePPM == newMaxSlippagePPM) {
+            return;
+        }
+
+        _maxSlippagePPM = newMaxSlippagePPM;
+
+        emit MaxSlippageUpdated({ prevMaxSlippagePPM: prevMaxSlippagePPM, newMaxSlippagePPM: newMaxSlippagePPM });
+    }
+
+    /**
+     * @dev set allowance to Uniswap V3 to the max amount if it's less than the input amount
+     */
+    function _setAllowance(Token token, uint256 inputAmount) private {
+        if (token.isNative()) {
+            return;
+        }
+        uint256 allowance = token.allowance(address(this), address(_uniswapV3Router));
+        if (allowance < inputAmount) {
+            // increase allowance to the max amount if allowance < inputAmount
+            token.safeIncreaseAllowance(address(_uniswapV3Router), type(uint256).max - allowance);
```

Comment:
If you'll upgrade OZ to 4.9.3, this will use the newer (and more gas efficient) `forceApprove`

Reply:
And if you wait for the version currently on their master branch (going to the next release), they have applied some additional improvements in the increase/decrease allowance functions.

Reply:
Estimations on the gas cost savings ?

Reply:
Basically, it should be almost the same. The main difference is that the newer approach has also a fallback to try approve(0) if the original approve fails (e.g., for USDT)